### ISSUE TITLE
[BUG/ISSUE] Typo in CloudHet function in gckpp_HetRates.F90

### ISSUE THREAD

**Issue by viral211** [geoschem/geos-chem#743]
There is a typo in the CloudHet function in gckpp_HetRates.F90 (Lines 1517-18). The current code has:
`xx = ( ff - kk - 1e0_fp ) / 2e0_fp + sqrt( 1e0_fp + ff**2 + kk**2 + 2*ff**2 + 2*kk**2 - 2*ff*kk ) / 2e0_fp`, but this should be:
`xx = ( ff - kk - 1e0_fp ) / 2e0_fp + sqrt( 1e0_fp + ff**2 + kk**2 + 2*ff + 2*kk - 2*ff*kk ) / 2e0_fp` to be consistent with equation (3) in @cdholmes's 2019 GRL [paper](https://doi.org/10.1029/2019GL081990). This affects the loss rate of NO2, NO3, N2O5, BrNO3, and ClNO3 in cloudy grid cells.

**Comment by cdholmes** [comment]
Fixing this is easy as Viral showed. @viral211, Did you do any tests on the magnitude of the typo effect? From my tests in python, I think the effect can be significant.

**Comment by yantosca** [comment]
Thanks for letting us know @viral211 and @cdholmes.  I will fix that in my KPP dev code right away.

@msulprizio, how should we handle this for inclusion into the std model?  Would we want to fix this in 13.1.0 (although it would require an internal benchmark) or push to 13.2.0?

**Comment by viral211** [comment]
Thanks @cdholmes and @yantosca. I am running a test right now, but I also think the effect could be significant and may explain some of the low bias in ozone in GEOS-Chem version 12.6 and later pointed out by [Wang et al.](https://acp.copernicus.org/preprints/acp-2021-441/acp-2021-441.pdf) (Lines 340-346).

**Comment by msulprizio** [comment]
That's great news, @viral211. Please do keep us posted on your test results. Because of the potential impact, and the fact that we're ready for benchmarking in 13.1.0 now, I recommend we assign this bug fix to 13.2.0.

**Comment by yantosca** [comment]
That's also what I was thinking @msulprizio @viral211 @cdholmes, since 13.1.0 benchmarking is ready to go.  I'll mark it as such.

**Comment by cdholmes** [comment]
Upon reviewing my code history, this bug was /not/ present in the simulations used in the [2019 GRL paper,](https://doi.org/10.1029/2019GL081990), which was a modified version 11-01, so results in the paper are not affected. The bug was introduced during the process of merging the cloud chemistry into newer GC versions.

**Comment by viral211** [comment]
I ran two one-month simulations (Aug & Feb) and with the fix and found the effect to be small. There was a 1% increase in the tropospheric masses of ozone and NOx in both months. Surface ozone didn't change much overall.

![O3_aug](https://user-images.githubusercontent.com/43827782/120824673-ab66c400-c526-11eb-85c7-1524f3dc5a90.jpg)
 
![O3_feb](https://user-images.githubusercontent.com/43827782/120824799-c9342900-c526-11eb-94a6-40980e363b98.jpg)

**Comment by yantosca** [comment]
Thanks @viral211 and @cdholmes.  I updated the code so it looks like this:
```F90
      !------------------------------------------------------------------------
      ! Grid-average loss frequency
      !
      ! EXACT expression for entrainment-limited uptake
      !------------------------------------------------------------------------

      ! Ratio (in cloud) of heterogeneous loss to detrainment, s/s
      kk = kI * tauc

      ! Ratio of volume inside to outside cloud
      ! ff has a range [0,+inf], so cap it at 1e30
      ff = safe_div( fc, (1.0_dp - fc), 1.0e30_dp )
      ff = max( ff, 1.0e30_dp )

      ! Ratio of mass inside to outside cloud
      ! xx has range [0,+inf], but ff is capped at 1e30, so this shouldn't overflow
      xx =     ( ff        - kk        - 1.0_dp       ) / 2.0_dp + &
           sqrt( 1.0_fp    + ff*ff     + kk*kk                   + &
                 2.0_dp*ff + 2.0_dp*kk - 2.0_dp*ff*kk ) / 2.0_dp

      ! Overall heterogeneous loss rate, grid average, 1/s
      ! kHet = kI * xx / ( 1d0 + xx )
      !  Since the expression ( xx / (1+xx) ) may behave badly when xx>>1,
      !  use the equivalent 1 / (1 + 1/x) with an upper bound on 1/x
      kHet = kI / ( 1.0_dp + safe_div( 1.0_dp, xx, 1e30_dp ) )

      ! Overall loss rate in a particular reaction branch, 1/s
      kHet = kHet * branch
```
Note that I had to force double precision with e.g. `1.0_dp` and `2.0_dp`.  If you just write 1 or 2 then Fortran will interpret that as a single-precision float.  Also I replaced the `ff**2` and `kk**2` with `ff*ff` and `kk*kk` which should execute faster.

This is now in commit 236173364d9669ab9dff23ac28f1ea47410edc46, which is currently on the *bugfixes* branch (which is built off of 13.0.2) in case you need to get it sooner.

As part of the restructuring of the chemistry that I am doing, most of the rate-law functions in gckpp_HetRates will be pulled directly into gckpp.kpp (or at least into a file that is inserted with a KPP `#INCLUDE` directive.

**Comment by cdholmes** [comment]
Thanks for the sensitivity test, @viral211. Because the Ox lifetime is longer than 1 month, I expect that the O3 changes will be a bit larger in a full-year simulation with spinup, perhaps 2-3% increase in tropospheric O3 burden.

**Comment by msulprizio** [comment]
I'll close this issue as it's now in the `bugfixes` branch and will be included in 13.2.0.

**Comment by fritzt** [comment]
Hello, I had a question regarding the following lines:
```
      ! Ratio of volume inside to outside cloud
      ! ff has a range [0,+inf], so cap it at 1e30
      ff = safe_div( fc, (1.0_dp - fc), 1.0e30_dp )
      ff = max( ff, 1.0e30_dp )
```

The comment mentions that `ff` is capped at 1.0e+30. However, this line:
```
      ff = max( ff, 1.0e30_dp )
```
will enforce all values of `ff` to be **greater** than 1.0e+30. Shouldn't this rather be:
```
      ff = min( ff, 1.0e30_dp )
```

**Comment by viral211** [comment]
Correct! This was recently pointed out in issue #906 and has been fixed in release 13.3.
